packages/hyperion-util/src/onNetworkRequest.ts

```diff
@@ -3,58 +3,80 @@
 import { assert } from "@hyperion/global";
 
-type NetworkRequest = {
+type NetworkRequestInfo = {
   readonly body?: RequestInit['body'],
   readonly method: string;
-  readonly url: string | URL; // TODO: should we always send on of the types?
+  readonly url: string;
 }
 
 /**
  * This is a generic event to be fired when a network request is about to happen.
  * Generally, various network api can be intercepted to provide a more unified way
  * of notifying the app about network requests.
  */
-export const onNetworkRequest = new Hook<(request: NetworkRequest) => void>();
-
-IWindow.fetch.onArgsObserverAdd((input, init) => {
-  let request: NetworkRequest;
-  if (typeof input === "string") {
-    request = {
-      body: init?.body,
-      method: init?.method ?? "get",
-      url: input,
-    };
-  } else if (input instanceof Request) {
-    request = {
-      body: input.body,
-      method: input.method,
-      url: input.url,
-    }
-  } else {
-    request = {
-      method: "GET",
-      url: input
-    }
-  }
+export const onNetworkRequestSend = new Hook<(request: XMLHttpRequest, requestInfo: NetworkRequestInfo) => void>();
 
-  onNetworkRequest.call(request);
-});
+/**
+ * This is a generic event to be fired when a network request has completed,
+ * whether successfully or unsuccessfully.
+ */
+export const onNetworkRequestLoad = new Hook<(request: XMLHttpRequest) => void>();
+
+// IWindow.fetch.onArgsObserverAdd((resource, options) => {
+//   let requestInfo: NetworkRequestInfo;
+//   if (resource instanceof Request) {
+//     requestInfo = {
+//       body: resource.body,
+//       method: resource.method,
+//       url: resource.url,
+//     }
+//   } else {
+//     const url = (typeof resource === "string")
+//       ? resource
+//       : resource.toString();
+//     requestInfo = {
+//       body: options?.body,
+//       method: options?.method ?? "get",
+//       url,
+//     }
+//   }
+
+//   onNetworkRequestSend.call(resource, requestInfo);
+// });
 
 //#region XHR
 const XHR_REQUEST_INFO_PROP = 'requestInfo';
 IXMLHttpRequest.open.onArgsObserverAdd(function (this, method, url) {
-  intercept.setVirtualPropertyValue<NetworkRequest>(this, XHR_REQUEST_INFO_PROP, { method, url });
+  intercept.setVirtualPropertyValue<NetworkRequestInfo>(
+    this,
+    XHR_REQUEST_INFO_PROP,
+    {
+      method,
+      url: url instanceof URL
+        ? url.href
+        : url,
+   });
 });
 
 IXMLHttpRequest.send.onArgsObserverAdd(function (this, body) {
-  const request = intercept.getVirtualPropertyValue<NetworkRequest>(this, XHR_REQUEST_INFO_PROP);
-  assert(request != null, `Unexpected situation! Request info is missing from xhr object`);
-  onNetworkRequest.call((body instanceof Document) ? request : {...request, body}); // assert already ensures request is not undefined
+  this.addEventListener('loadend', (event) => {
+    if (event.target instanceof XMLHttpRequest) {
+      onNetworkRequestLoad.call(event.target);
```

Comment:
Could we just read the response and send it in the event instead of the object itself?